ssl/ssl_key_share.cc

```diff
@@ -294,6 +295,233 @@ class CECPQ2KeyShare : public SSLKeyShare {
+      EVP_PKEY_kem_new_raw_public_key(nid_, peer_key.begin(), peer_key.size()));
+
+    if (!pkey) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Initialize the context with the pkey
+    ctx_.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len_,
+                              nullptr, &secret_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // EVP_PKEY_encapsulate() will generate a shared secret, then encrypt it
+    // with the peer's public key. We send the resultant ciphertext to the
+    // peer by writing it to |out_public_key|, then...
+    Array<uint8_t> ciphertext;
+    Array<uint8_t> shared_secret;
+    if (!ciphertext.Init(ciphertext_len_) || !shared_secret.Init(secret_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t ciphertext_bytes_written = ciphertext_len_;
+    size_t secret_bytes_written = secret_len_;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), ciphertext.data(),
+                              &ciphertext_bytes_written, shared_secret.data(),
+                              &secret_bytes_written) ||
+                              ciphertext_bytes_written != ciphertext_len_ ||
+                              secret_bytes_written != secret_len_ ||
+                              !CBB_add_bytes(out_public_key, ciphertext.data(), ciphertext_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // ...we retain the shared secret for our own use.
+    *out_secret = std::move(shared_secret);
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    // Basic nullptr checks
+    if (!out_alert) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // Set alert to internal error by default
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if (!out_secret) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should have been initialized by a previous call to Offer()
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len_, nullptr,
+                              &secret_len_)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Ensure that peer_key is valid
+    if (!peer_key.data() || peer_key.size() != ciphertext_len_) {
```

Comment:
the naming is very confusing, for example `peer_key.size() != ciphertext_len_`

Reply:
Added clarifying comment to the beginning of the function